command/head.go

```diff
@@ -0,0 +1,206 @@
+		Usage:    "print remote object metadata",
+
+		CustomHelpTemplate: headHelpTemplate,
+
+		Flags: []cli.Flag{
+			&cli.StringFlag{
+				Name:  "version-id",
+				Usage: "use the specified version of an object",
+			},
+			&cli.BoolFlag{
+				Name:  "raw",
+				Usage: "disable the wildcard operations, useful with filenames that contains glob characters",
+			},
+		},
+
+		Before: func(c *cli.Context) error {
+			err := validateHeadCommand(c)
+			if err != nil {
+				printError(commandFromContext(c), c.Command.Name, err)
+			}
+			return err
+		},
+		Action: func(c *cli.Context) (err error) {
+			defer stat.Collect(c.Command.FullName(), &err)()
+
+			op := c.Command.Name
+			fullCommand := commandFromContext(c)
+			src, err := url.New(c.Args().Get(0), url.WithVersion(c.String("version-id")),
+				url.WithRaw(c.Bool("raw")))
+			if err != nil {
+				printError(fullCommand, op, err)
+				return err
+			}
+
+			return Head{
+				src:         src,
+				op:          op,
+				fullCommand: fullCommand,
+				storageOpts: NewStorageOpts(c),
+			}.Run(c.Context)
+		},
+	}
+	cmd.BashComplete = getBashCompleteFn(cmd, true, false)
+	return cmd
+}
+
+type Head struct {
+	src         *url.URL
+	op          string
+	fullCommand string
+	storageOpts storage.Options
+}
+
+func (h Head) Run(ctx context.Context) error {
+	client, err := storage.NewRemoteClient(ctx, h.src, h.storageOpts)
+	if err != nil {
+		printError(h.fullCommand, h.op, err)
+		return err
+	}
+
+	if h.src.IsBucket() {
+		err := client.HeadBucket(ctx, h.src)
+		if err != nil {
+			printError(h.fullCommand, h.op, err)
+			return err
+		}
+
+		msg := HeadBucketMessage{
+			Key: h.src.String(),
+		}
+
+		log.Info(msg)
+
+		return nil
+	}
+
+	object, metadata, err := client.HeadObject(ctx, h.src)
+	if err != nil {
+		printError(h.fullCommand, h.op, err)
+		return err
+	}
+
+	msg := HeadObjectMessage{
+		Key:                  object.URL.String(),
+		ContentType:          metadata.ContentType,
+		ServerSideEncryption: metadata.EncryptionMethod,
+		LastModified:         object.ModTime,
+		ContentLength:        object.Size,
+		StorageClass:         string(object.StorageClass),
+		VersionID:            object.VersionID,
+		ETag:                 object.Etag,
+		Metadata:             metadata.UserDefined,
+	}
+
+	log.Info(msg)
+
+	return nil
+}
+
+type HeadObjectMessage struct {
+	Key                  string            `json:"Key,omitempty"`
+	ContentType          string            `json:"ContentType,omitempty"`
+	ServerSideEncryption string            `json:"ServerSideEncryption,omitempty"`
+	LastModified         *time.Time        `json:"LastModified,omitempty"`
+	ContentLength        int64             `json:"ContentLength,omitempty"`
+	StorageClass         string            `json:"StorageClass,omitempty"`
+	VersionID            string            `json:"VersionID,omitempty"`
+	ETag                 string            `json:"ETag,omitempty"`
+	Metadata             map[string]string `json:"Metadata"`
```

Comment:
let's keep json keys consistent across the project;  snake case
https://github.com/peak/s5cmd/blob/89218dc7fd484030367b3139961ba1be061d8b2b/storage/storage.go#L111-L122

Reply:
Some of the field names have not been defined yet.

- `type` not same as `ContentType`
- there is no json keys for `ServerSideEncryption` and `Metadata` in the project.

My suggestions:
- `ContentType -> "content_type"`
- `Metadata -> "metadata"`
- `ServerSideEncryption -> "server_side_encryption"`

Reply:
i've shared the above as an example, you can just use snake case variants of the field names, nothing to decide IMHO.

Reply:
done.